packages/runtime/container-runtime/src/containerRuntime.ts

```diff
@@ -1752,30 +1758,61 @@ export class ContainerRuntime extends TypedEventEmitter<IContainerRuntimeEvents>
         assert(this._orderSequentiallyCalls === 0,
             0x24c /* "Cannot call `flush()` from `orderSequentially`'s callback" */);
 
-        if (!this.deltaSender) {
-            return;
-        }
+        const batch = this.batchManager.popBatch();
 
-        // Let the PendingStateManager know that there was an attempt to flush messages.
-        // Note that this should happen before the `this.needsFlush` check below because in the scenario where we are
-        // not connected, `this.needsFlush` will be false but the PendingStateManager might have pending messages and
-        // hence needs to track this.
-        this.pendingStateManager.onFlush();
+        const length = batch.length;
 
-        // If flush has already been called then exit early
-        if (!this.needsFlush) {
-            return;
+        if (length > 1) {
+            batch[0].metadata = { ...batch[0].metadata, batch: true };
+            batch[length - 1].metadata = { ...batch[length - 1].metadata, batch: false };
         }
 
-        this.needsFlush = false;
+        const refSeqNumber = this.deltaManager.lastSequenceNumber;
+        let clientSequenceNumber: number = -1;
 
         // Did we disconnect in the middle of turn-based batch?
         // If so, do nothing, as pending state manager will resubmit it correctly on reconnect.
-        if (!this.canSendOps()) {
-            return;
+        if (this.canSendOps()) {
+            if (this.context.submitBatchFn !== undefined) {
+                const batchToSend: IBatchMessage[] = [];
+                for (const message of batch) {
+                    batchToSend.push({ contents: message.contents, metadata: message.metadata });
+                }
+                clientSequenceNumber = this.context.submitBatchFn(batchToSend);
+            } else {
+                // Legacy path - supporting old loader versions. Can be removed only when LTS moves above
+                // version that has support for batches (submitBatchFn)
+                for (const message of batch) {
+                    clientSequenceNumber = this.context.submitFn(
+                        MessageType.Operation,
+                        message.deserializedContent,
+                        true, // batch
+                        message.metadata);
+                }
+
+                this.deltaSender.flush();
+            }
         }
 
-        return this.deltaSender.flush();
+        clientSequenceNumber -= batch.length - 1;
+
+        // Let the PendingStateManager know that a message was submitted.
+        // In future, need to shift toward keeping batch as a whole!
+        for (const message of batch) {
+            this.pendingStateManager.onSubmitMessage(
+                message.deserializedContent.type,
+                clientSequenceNumber,
+                refSeqNumber,
```

Comment:
I am not familiar with the code that uses this refSeqNumber (stashing) but wouldn't it cause a problem now that a bunch of messages will have the same refSeqNumber?

Reply:
I had same question :) It used for stashed op workflow (see Container.closeAndGetPendingLocalState()) to reason when to replay these ops.
There is no change in behavior here - a batch can be produced only by synchronous code and can't be interrupted by processing remote ops. So, all ops in a batch always have same reference sequence number. We can probably add here some asserts to be more defensive in this space - opened https://dev.azure.com/fluidframework/internal/_workitems/edit/1848 to follow up